themes/src/main/resources/theme/base/admin/resources/js/authz/authz-controller.js

```diff
@@ -1213,90 +1213,129 @@ module.controller('ResourceServerPolicyScopeDetailCtrl', function($scope, $route
                     });
                 }
             }
+            $scope.applyToResourceType = function() {
+                if ($scope.applyToResourceTypeFlag) {
+                    $scope.selectedResource = null;
+                } else {
+                    $scope.policy.resourceType = null;
+                }
+                $scope.selectedScopes = [];
+                $scope.changed = true;
+            }
         },
 
         onInitUpdate : function(policy) {
-            ResourceServerPolicy.resources({
-                realm : $route.current.params.realm,
-                client : client.id,
-                id : policy.id
-            }, function(resources) {
-                if (resources.length > 0) {
-                    for (i = 0; i < resources.length; i++) {
-                        ResourceServerResource.get({
-                            realm: $route.current.params.realm,
-                            client: client.id,
-                            rsrid: resources[0]._id,
-                        }, function (resource) {
-                            ResourceServerResource.query({
+            if (!policy.resourceType) {
```

Comment:
Instead of adding a long if-then-else clause, please break this code up into at least two new functions:

```
if (!policy.resourceType) {
   functionA();
} else {
   functionB();
}
```

Reply:
> This isn't working quite right:
> 
> 1. Turn `Apply to Resource Type` to On
> 2. Specify resource type
> 3. Specify a scope.
> 4. Press save
> 5. Turn `Apply to Resource Type` to Off
> 6. Scope from step 3 goes away (is that correct?)
> 7. Specify scope again
> 8. Press save
> 9. `Apply to Resource Type` has turned back on by itself

This was the problem in the existing code for "Resource" permissions also.  When the same steps are performed in "Resource" Permission form, the issue is produced.

It has been fixed with the following code. Whenever flag for "resource type" is toggled, values are adjusted accordingly.

```
$scope.applyToResourceType = function() {
    // if previously apply to resource type flag is selected,
    // assume that it will be disabled now and accordingly, 
    // set values for selectedResource or resourceType
    const prevApplyToResourceTypeFlag = $scope.applyToResourceTypeFlag;
    if (!prevApplyToResourceTypeFlag) {
        $scope.selectedResource = null;
    } else {
        $scope.policy.resourceType = null;
    }
    if (ResourceServerScope != null) {
        $scope.selectedScopes = [];
        $scope.changed = true;
    }
}
```

Reply:
> Instead of adding a long if-then-else clause, please break this code up into at least two new functions:
> 
> ```
> if (!policy.resourceType) {
>    functionA();
> } else {
>    functionB();
> }
> ```

Now, this is really interesting one. 😃 

I have created common method (ResourceServerPolicyCommonHandler) to use between "ResourceServerPolicyResourceDetailCtrl" and "ResourceServerPolicyScopeDetailCtrl". 

This way, significant code has been reduced.

To review the code effectively, I would suggest that instead of PR review (where proper block changes are not displayed), please compare ResourceServerPolicyCommonHandler method code with existing ResourceServerPolicyResourceDetailCtrl and ResourceServerPolicyScopeDetailCtrl.

Pointers about ResourceServerPolicyCommonHandler:

1. Instead of ResourceServerPolicy, have used ResourceServerPermission, as permission class is derived from policy only in backend and from a usage point of view, that is consistent.
2. For scope-based permission, ResourceServerScope will be passed otherwise null. In the code, wherever scope needs to be handled, checking for ResourceServerScope is used.
3. In UI, resource-server-policy-scope-detail.html file, instead of using two different controls for selectedScope, now single control is used.
4. As single control is used for selected scopes now, to handle scenarios about getting scopes from scopes of the selected resource (resourceScope), logic is added in $scope.scopesUiSelect
5. Due to this change, code for onInitUpdate is simple now, where earlier "scopes[i].id" was pushed in $scope.selectedScopes in some cases and sometimes "scopes[i]" was pushed.
6. This way, one more issue is solved that exists in the current code. (Explained Below)

The issue which is fixed now: On changing of resource, scopes are cleared in models but display is not updated.

1. Open "Scope" Permission.
2. Select Resource "Resource A".
3. Select Scope "Scope1".
4. Clear the resource.
5. Scope will be cleared.
6. Select Resource "Resource B". (Here assume that "Scope1" is not mapped with "Resource B").
7. Still "Scope1" will be displayed. It looks like "select2" refresh issue.
8. While trying to save this, it will give an error. It means, internally it uses proper data model only where selectedScopes are empty.


I understand that it is a major change from a UI perspective now, but I thought this is a good opportunity to clean up the code and it has brought consistency for both types of permissions handling.

Reply:
@pedroigor I think, you will also need to review changes as I have changed some of the calling services. Earlier ResourceServerResource and ResourceServerPolicy were used. Instead of that now I am using ResourceServerPermission for that. Although I have studied backend code and also checked behavior, your review would be helpful here. 

@ssilvert, please let me know your feedback also.